src/TestSuite/LogTestTrait.php

```diff
@@ -0,0 +1,126 @@
+<?php
+declare(strict_types=1);
+
+/**
+ * CakePHP(tm) : Rapid Development Framework (https://cakephp.org)
+ * Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ *
+ * Licensed under The MIT License
+ * For full copyright and license information, please see the LICENSE.txt
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ * @link          https://cakephp.org CakePHP(tm) Project
+ * @since         5.1.0
+ * @license       https://opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Cake\TestSuite;
+
+use Cake\Log\Engine\ArrayLog;
+use Cake\Log\Log;
+use InvalidArgumentException;
+
+/**
+ * Make assertions on logs
+ */
+trait LogTestTrait
+{
+    /**
+     * Reset log configs
+     *
+     * @after
+     * @return void
+     */
+    public function cleanupLog(): void
+    {
+        Log::reset();
+    }
+
+    /**
+     * @param array|string $engines The engine(s) which should receive a log message
+     * @return void
+     */
+    public function setupLog(array|string $engines): void
+    {
+        Log::reset();
+        $engines = (array)$engines;
+        foreach ($engines as $engineName => $engineConfig) {
+            if (is_string($engineConfig)) {
+                Log::setConfig($engineConfig, ['className' => 'Array']);
+            } else {
+                $engineConfig['className'] = 'Array';
+                Log::setConfig($engineName, $engineConfig);
+            }
+        }
+    }
+
+    /**
+     * @param string $engine The engine which should receive a log message
+     * @param string $expectedMessage The message which should be inside the log engine
+     * @param string|null $level The level of the expected message
+     * @param string $failMsg The error message if the message was not in the log engine
+     * @return void
+     */
+    public function assertLogMessage(
+        string $engine,
+        string $expectedMessage,
+        ?string $level = null,
+        string $failMsg = ''
+    ): void {
```

Comment:
Is the engine a log message was recorded on the most important part of the log message? To me the goal of testing logs is to:

1. Validate that a given log message was created.
2. That the log message was emit at the right level,
3. That the log has the right scopes so that I can be confident that logs will be routed correctly.

The engine that captured the log message is less important to me than the level. What do you think about the assertion signatures being

```php
assertLogMessage(string $level, string $message, ?string $engine = null, string $failMessage = '')
```

I can put together a commit/diff for the change I have in mind.

Reply:
sure, go ahead